smart_open/bytebuffer.py

```diff
@@ -159,3 +159,22 @@ def fill(self, source, size=-1):
 
         self._bytes += new_bytes
         return len(new_bytes)
+
+    def readline(self, terminator):
+        """Read a line from this buffer efficiently.
+
+        A line is a contiguous sequence of bytes that ends with either:
+
+        1. The ``terminator`` character
```

Comment:
How does this play with Python's "universal newline" functionality? Is it fully compatible?

IIRC Python considers multiple byte terminators as possible line ending: `\n`, `\r`, `\r\n`.

Reply:
We don't support universal newlines at all ().

```python
    if mode not in ('rb', 'rb+', 'wb', 'wb+', 'ab', 'ab+'):
        #
        # This should really be a ValueError, but for the sake of compatibility
        # with older versions, which raise NotImplementedError, we do the same.
        #
        raise NotImplementedError('unsupported mode: %r' % mode)
```